lib/util/extractSourceMap.js

```diff
@@ -0,0 +1,414 @@
+
+			return { sourceURL: "", sourceContent };
+		}
+
+		if (skipReading) {
+			return { sourceURL: url };
+		}
+
+		if (protocol === "file:") {
+			const pathFromURL = urlUtils.fileURLToPath(url);
+			const sourceURL = path.normalize(pathFromURL);
+			const { data: sourceContent } = await fetchFromFilesystem(fs, sourceURL);
+
+			return { sourceURL, sourceContent };
+		}
+
+		throw new Error(
+			`Failed to parse source map: '${url}' URL is not supported`
+		);
+	}
+
+	// 2. It's a scheme-relative
+	if (/^\/\//.test(url)) {
+		throw new Error(
+			`Failed to parse source map: '${url}' URL is not supported`
+		);
+	}
+
+	// 3. Absolute path
+	if (isAbsolute(url)) {
+		let sourceURL = path.normalize(url);
+
+		let sourceContent;
+
+		if (!skipReading) {
+			const possibleRequests = [sourceURL];
+
+			if (url.startsWith("/")) {
+				possibleRequests.push(
+					getAbsolutePath(fs, context, sourceURL.slice(1), sourceRoot || "")
+				);
+			}
+
+			const result = await fetchPathsFromFilesystem(fs, possibleRequests);
+
+			sourceURL = result.path;
+			sourceContent = result.data;
+		}
+
+		return { sourceURL, sourceContent };
+	}
+
+	// 4. Relative path
+	const sourceURL = getAbsolutePath(fs, context, url, sourceRoot || "");
+
+	let sourceContent;
+
+	if (!skipReading) {
+		const { data } = await fetchFromFilesystem(fs, sourceURL);
+
+		sourceContent = data;
+	}
+
+	return { sourceURL, sourceContent };
+}
+
+/**
+ * Extract source map from code content
+ * @param {string | Buffer<ArrayBufferLike>} stringOrBuffer The input code content as string or buffer
+ * @param {string} resourcePath The path to the resource file
+ * @param {InputFileSystem} fs The file system interface for reading files
+ * @param {((sourceMappingURL: string, resourcePath: string) => string)=} filterSourceMappingUrl Optional filter function for source mapping URLs
+ * @returns {Promise<{source: string | Buffer<ArrayBufferLike>, sourceMap: string | RawSourceMap | undefined, fileDependencies: string[]}>} Promise resolving to extracted source map information
+ */
+async function extractSourceMap(
+	stringOrBuffer,
+	resourcePath,
+	fs,
+	filterSourceMappingUrl
+) {
+	const input =
+		typeof stringOrBuffer === "string"
+			? stringOrBuffer
+			: stringOrBuffer.toString("utf8");
+	const inputSourceMap = undefined;
+	const output = {
+		source: stringOrBuffer,
+		sourceMap: inputSourceMap,
+		fileDependencies: /** @type {string[]} */ ([])
+	};
+	const { sourceMappingURL, replacementString } = getSourceMappingURL(input);
+
+	if (!sourceMappingURL) {
+		return output;
+	}
+
+	const behaviourSourceMappingUrl =
+		typeof filterSourceMappingUrl === "function"
+			? filterSourceMappingUrl(sourceMappingURL, resourcePath)
+			: "consume";
+
+	switch (behaviourSourceMappingUrl) {
+		case "skip":
+			return output;
+		case "false":
+		case "remove":
+			return {
+				source: input.replace(replacementString, ""),
+				sourceMap: inputSourceMap,
+				fileDependencies: []
+			};
+	}
```

Comment:
I think we don't need it here and we can be always use `consume`. Why?

You can set `extractSourceMap: false` using rules, for example:

```
{
		target: "node",
		entry: "./extract2",
		devtool: "source-map",
		module: {
			rules: [
				{
                    test: /\.js$/,
					extractSourceMap: true
				},
                {
                    test: /keep-source-map-comment\.js$/,
					extractSourceMap: false
				}
			]
		}
	},
```

So `skip` are unnecessary, `false` and `remove` is out of scope, I think if you really need to remove source map comments you can write a simple loader to do it, it was implemented a long time ago and looks like no one using it